implementation/src/main/java/io/smallrye/config/SecretKeysConfigSourceInterceptor.java

```diff
@@ -1,27 +1,72 @@
 package io.smallrye.config;
 
+import static io.smallrye.common.expression.Expression.Flag.DOUBLE_COLON;
+import static io.smallrye.common.expression.Expression.Flag.LENIENT_SYNTAX;
+import static io.smallrye.common.expression.Expression.Flag.NO_SMART_BRACES;
+import static io.smallrye.common.expression.Expression.Flag.NO_TRIM;
+
+import java.util.Collections;
 import java.util.HashSet;
 import java.util.Iterator;
+import java.util.Map;
 import java.util.Set;
+import java.util.function.BiConsumer;
 
 import javax.annotation.Priority;
 
+import io.smallrye.common.expression.Expression;
+import io.smallrye.common.expression.ResolveContext;
+
 @Priority(Priorities.LIBRARY + 100)
 public class SecretKeysConfigSourceInterceptor implements ConfigSourceInterceptor {
     private static final long serialVersionUID = 7291982039729980590L;
 
     private final Set<String> secrets;
+    private final Map<String, SecretKeysHandler> handlers;
 
     public SecretKeysConfigSourceInterceptor(final Set<String> secrets) {
+        this(secrets, Collections.emptyMap());
+    }
+
+    public SecretKeysConfigSourceInterceptor(final Set<String> secrets, final Map<String, SecretKeysHandler> handlers) {
         this.secrets = secrets;
+        this.handlers = handlers;
     }
 
     @Override
     public ConfigValue getValue(final ConfigSourceInterceptorContext context, final String name) {
         if (SecretKeys.isLocked() && isSecret(name)) {
             throw ConfigMessages.msg.notAllowed(name);
         }
-        return context.proceed(name);
+        ConfigValue configValue = context.proceed(name);
+
+        if (configValue == null || handlers.isEmpty()) {
+            return configValue;
+        }
+
+        // This executes before the Expression resolution.
+        // If we find a handler we execute it, if not we proceed
+        Expression expression = Expression.compile(configValue.getValue(), LENIENT_SYNTAX, NO_TRIM, NO_SMART_BRACES,
+                DOUBLE_COLON);
```

Comment:
Really the point of using the expression compiler is that we can combine expression resolution with double-colon expansion in a single step. Once we have two separate resolvers, then things get messy including recursive expression support.

What about adding a `resolveSecret(String name, String value)` to `ConfigSourceInterceptorContext`, and adding secret resolvers to the base implementation along with converters and config sources?

Reply:
> Really the point of using the expression compiler is that we can combine expression resolution with double-colon expansion in a single step. Once we have two separate resolvers, then things get messy including recursive expression support.

Maybe I'm missing something, but when I've tried with a single expression compiler and the `DOUBLE_COLON` flag, the entire expression becomes part of the key when the expression contains a single colon, meaning that it would break the default values expressions. Maybe that is a bug from what it is intended? :) 

> What about adding a `resolveSecret(String name, String value)` to `ConfigSourceInterceptorContext`, and adding secret resolvers to the base implementation along with converters and config sources?

The idea was to add the `SecretKeysHandler` (we can rename it to `SecretKeysResolver`) in the base implementation. Right now, they are just discovered via ServiceLoader. We can add APIs to directly access them, I was trying to be minimalist for now. 

How do you intend to use `resolveSecret(String name, String value)`? Expose it as a separate API in `SmallRyeConfig` to look up secrets?

Reply:
> > Really the point of using the expression compiler is that we can combine expression resolution with double-colon expansion in a single step. Once we have two separate resolvers, then things get messy including recursive expression support.
> 
> Maybe I'm missing something, but when I've tried with a single expression compiler and the `DOUBLE_COLON` flag, the entire expression becomes part of the key when the expression contains a single colon, meaning that it would break the default values expressions. Maybe that is a bug from what it is intended? :)

You mean when the expression is e.g. `foo:bar::baz:zap` or for normal expressions as well?

> > What about adding a `resolveSecret(String name, String value)` to `ConfigSourceInterceptorContext`, and adding secret resolvers to the base implementation along with converters and config sources?
> 
> The idea was to add the `SecretKeysHandler` (we can rename it to `SecretKeysResolver`) in the base implementation. Right now, they are just discovered via ServiceLoader. We can add APIs to directly access them, I was trying to be minimalist for now.
> 
> How do you intend to use `resolveSecret(String name, String value)`? Expose it as a separate API in `SmallRyeConfig` to look up secrets?

No, it would be usable by interceptors (specifically the expression interceptor, to implement the `::` lookup) but otherwise not accessible from the main API. The implementation(s) would be loaded via `ServiceLoader` as you already have it.

Reply:
The interceptor context is AFAIK the only thing that would be available to the expression interceptor that would allow it to interact with the greater configuration system (outside of hacky things like global variables or thread-locals or things like that). That said, we could opt to make the method non-public.

Reply:
We do initialize the `ExpressionConfigSourceInterceptor` in the builder. What if we just discover and populate the `SecretKeysResolvers` directly into that interceptor? If we need them later for outside of the expression interceptor we can rewrite the implementation without impact.

Reply:
Sure, that seems reasonable. :+1: